### PR TITLE
@craigspaeth: User authentication; BidderPositions... etcetera

### PR DESCRIPTION
**Pull request by dzucconi**
I want to be able to fetch user authenticated stuff along-side non-authenticated stuff that's cached.

I'm not that happy with how I resolved this but it does the job for now.

Since we don't leverage caching on any unauthenticated stuff then this will get slow if you try to orchestrate a ton of things, so beware.

yields:

```
# Cached
info Loading: sale/los-angeles-modern-auctions-march-2015...
info Elapsed: 0s 5.736123ms - sale/los-angeles-modern-auc...
info Loading: sale/los-angeles-modern-auctions-march-2015...
info Elapsed: 0s 6.056311ms - sale/los-angeles-modern-auc...

# Uncached
info Loading: me/bidder_positions?artwork_id=ed-ruscha-co...
info Elapsed: 0s 266.647497ms - me/bidder_positions?artwo...

```

### CODE REVIEW

**Comment by dzucconi** [comment]
Everything prior to 1d892a8fb576db13c1edb458f71f07ba30f0ff46 is random refactoring; slipped that in

**Comment by dzucconi** [comment]
I can't for the life of me figure out how to get back colored logs within Express... any ideas?

File: server.js [hunk 1]
```diff
@@ -46,13 +47,20 @@ app.get('/favicon.ico', (req, res) => {
 
 app.all('/graphql', (req, res) => res.redirect('/'));
 
-app.use('/', cors(), morgan('combined'), graphqlHTTP(() => {
+app.use('/', cors(), morgan('combined'), graphqlHTTP(request => {
+  info('----------');
+
   loaders.clearAll();
 
+  const accessToken = request.headers['x-access-token'];
```
Comment (dzucconi) [inline server.js]: I've been using https://github.com/skevy/graphiql-app which will let you set headers (but is pretty buggy). Force is already sending this header.

**Comment by dzucconi** [comment]
bump @craigspaeth

**Comment by craigspaeth** [comment]
Super exciting! Sorry, digging in now.

File: schema/bidder_position.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import date from './fields/date';
+import {
+  GraphQLInt,
+  GraphQLBoolean,
+  GraphQLString,
+  GraphQLObjectType,
+} from 'graphql';
+
+const BidderPositionType = new GraphQLObjectType({
```
Comment (craigspaeth) [inline schema/bidder_position.js]: W00t W00t! Thanks for writing this!

File: schema/me.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+      resolve: ({ default_profile_id }) =>
+        gravity(`profile/${default_profile_id}`),
+    },
+  },
+});
+
+export default {
+  type: Me,
+  resolve: (root, options, { rootValue: { accessToken } }) => {
+    return gravity.with(accessToken)('me');
```
Comment (craigspaeth) [inline schema/me.js]: Neat API here! `gravity.with()()` :+1:

File: schema/me.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      type: Profile.type,
+      resolve: ({ default_profile_id }) =>
+        gravity(`profile/${default_profile_id}`),
+    },
+  },
+});
+
+export default {
+  type: Me,
+  resolve: (root, options, { rootValue: { accessToken } }) => {
```
Comment (dzucconi) [inline schema/me.js]: So yah—we can pass whatever we want down in the `rootValue` on a per-request basis, and that gets included as the third arg on every `resolve`. So we just pull out the `accessToken` and authenticate the gravity call with it. I think we could pass a single preauthenticated loader down too; but that can be a later refactoring; I was having a little difficulty figuring out how to safely clear it, so this seemed like the simplest first step

File: schema/sale_artwork.js [hunk 5]
```diff
@@ -33,6 +40,17 @@ const SaleArtworkType = new GraphQLObjectType({
...
         type: GraphQLInt,
       },
+      bidder_positions: {
+        type: new GraphQLList(BidderPosition.type),
+        resolve: (sale_artwork, options, { rootValue: { accessToken } }) => {
+          return gravity.with(accessToken)('me/bidder_positions', {
+            artwork_id: sale_artwork.id,
+            sale_id: sale_artwork.sale_id,
+          })
+            .then(response => enhance(response, { sale_artwork }))
```
Comment (dzucconi) [inline schema/sale_artwork.js]: We're going to need access to the `sale_artwork` for a few different fields that I haven't implemented yet

File: schema/me.js (hunk 4 above)
Comment (craigspaeth) [inline schema/me.js]: So when composing a GraphQL you pass the `accessToken` in a `rootValue` param? Or is it passed through the headers? Trying to wrap my head around what this looks like from the client-perpective.

Comment (dzucconi) [inline schema/me.js]: So this is passed automatically from Force: https://github.com/artsy/force/blob/8c53c0889890af9c84fa646451c5cfb64a9211da/lib/metaphysics.coffee#L9-L10

You won't have to do anything on that end. When this server gets the request it just pulls it out and pushes it onto the rootValue: https://github.com/dzucconi/metaphysics/commit/1d892a8fb576db13c1edb458f71f07ba30f0ff46#diff-78c12f5adc1848d13b1c6f07055d996eR55

Comment (craigspaeth) [inline schema/me.js]: Ah cool :+1:

**Comment by craigspaeth** [comment]
Looks great! I think it'll be interesting to see how we solve the caching challenge with user authenticated endpoints. I definitely support not tackling that now and leaving it uncached though. One thing to keep in mind when we do cross that bridge—considering endpoints that return different responses based on role, e.g. the `:all` json vs `:public` json. Naively my gut would say to just add the user's id to the cache key... but there's only [two hard things](http://martinfowler.com/bliki/TwoHardThings.html) in computer science.

**Pull request merged**